src/target/riscv/riscv-013.c

```diff
@@ -1784,6 +1780,37 @@ static void deinit_target(struct target *target)
 	info->version_specific = NULL;
 }
 
+static int smp_add_ext_trigger(struct target *target, unsigned int group,
+		riscv_ext_trigger_t external_trigger)
+{
+	uint32_t write_val = DM_DMCS2_HGSELECT;
+	assert(group <= 31);
+	assert(external_trigger.dmexttrigger < 16);
+	write_val = set_field(write_val, DM_DMCS2_GROUP, group);
+	write_val = set_field(write_val, DM_DMCS2_GROUPTYPE, external_trigger.grouptype);
+	write_val = set_field(write_val, DM_DMCS2_DMEXTTRIGGER, external_trigger.dmexttrigger);
+	if (dm_write(target, DM_DMCS2, write_val) != ERROR_OK)
+		return ERROR_FAIL;
+	write_val = set_field(write_val, DM_DMCS2_HGWRITE, 1);
+	if (dm_write(target, DM_DMCS2, write_val) != ERROR_OK)
+		return ERROR_FAIL;
+
+	uint32_t read_val;
+	if (dm_read(target, &read_val, DM_DMCS2) != ERROR_OK)
+		return ERROR_FAIL;
+	if (get_field(read_val, DM_DMCS2_GROUP) == group &&
+		get_field(read_val, DM_DMCS2_DMEXTTRIGGER) == external_trigger.dmexttrigger &&
+		get_field(read_val, DM_DMCS2_HGSELECT) == 1) {
+		LOG_TARGET_INFO(target, "External trigger %d added to %s group %d", external_trigger.dmexttrigger,
+				external_trigger.grouptype ? "resume" : "halt", group);
+	} else {
+		LOG_TARGET_ERROR(target, "External trigger %d not supported %s group %d", external_trigger.dmexttrigger,
+				external_trigger.grouptype ? "resume" : "halt", group);
```

Comment:
Suggested reword for clarity:

```suggestion
		LOG_TARGET_ERROR(target, "Could not add external trigger %d to %s group %d", external_trigger.dmexttrigger,
				external_trigger.grouptype ? "resume" : "halt", group);
```

Reply:
Addressed